### PR TITLE
Add helpers to create post control buttons

### PR DESCRIPTION
**Pull request by invalidCards**
#### User-facing changes
- None.

#### Technical explanation
Allows developers to easily create post control buttons using the RemixIcon library that is already used throughout the extension. This includes:
- A dev script (run with Node.js) to create a JSON lookup of the SVG paths for all RemixIcon icons;
- Helper functions added to `interface.js` to look up SVG paths, create button templates, and clone the templates while adding a callback to the `click` event at the same time; 
- A global stylesheet that copies the styles used by Tumblr vanilla buttons in order to blend in as much as possible.

#### Issues this closes
None, but it is a card on the To-Do list of the Alpha project.

### CODE REVIEW

File: src/util/interface.js [hunk 1]
```diff
@@ -84,5 +84,65 @@
...
+
+  /**
+   * Create a button template that can be cloned with cloneFooterButton() for inserting into the controls of a post.
+   *
+   * @param {string} iconClass The RemixIcon class of the requested icon
+   * @param {string} buttonClass An extra class to identify the extension that added the button
+   * @returns {HTMLDivElement} A button that can be cloned with cloneFooterButton()
+   */
+  const getPostFooterButton = async function (iconClass, buttonClass) {
+    const queueButtonContainer = document.createElement('div');
```
Comment (AprilSylph) [inline src/util/interface.js:111]: this and `queueButtonInner` should definitely have more generic names :P

**Review by AprilSylph — CHANGES_REQUESTED** [review]
i think i'd prefer if the code consistently called them "control buttons" or "control icons" to align with Tumblr's terminology

also, i'm preferring snake_case for filenames we create... i should probably note that in `CONTRIBUTING.md` somehow

- ~~`dev/remix_svg_to_json.js`~~
- `src/footer_button.css` (actually would prefer `src/control_buttons.css` or `src/control_icons.css`)
- `src/util/remixicon_svg.json`

File: src/util/interface.js (hunk 1 above)
Comment (invalidCards) [reply src/util/interface.js:111]: Whoops, I just C/P'd it from my old Quick Queue PR without looking >.<

**Review by AprilSylph — CHANGES_REQUESTED** [review]
code otherwise LGTM, just a couple things:
- for jsdoc `@param` tags, i prefer a hyphen to separate the description from the parameter name
- i feel like `remixicon_svg.json` should be in `lib/` since it's not really by us

File: src/control_buttons.css [hunk 2]
```diff
@@ -0,0 +1,25 @@
+.xkit-post-button-container {
```
Comment (AprilSylph) [inline src/control_buttons.css:1]:
```suggestion
.xkit-control-button-container {
```

File: src/control_buttons.css [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  font-family: var(--font-family);
+  font-size: 1rem;
+  font-weight: 700;
+  margin-left: 20px;
+  position: relative;
+  line-height: 1;
+}
+
+.xkit-post-button-container-span {
```
Comment (AprilSylph) [inline src/control_buttons.css:10]:
```suggestion
.xkit-control-button-container-span {
```

File: src/control_buttons.css [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+  vertical-align: top;
+  display: inline-block;
+  width: inherit;
+}
+
+.xkit-post-button {
```
Comment (AprilSylph) [inline src/control_buttons.css:16]:
```suggestion
.xkit-control-button {
```

File: src/control_buttons.css [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+  color: inherit;
+  font: inherit;
+}
+
+.xkit-post-button-inner {
```
Comment (AprilSylph) [inline src/control_buttons.css:21]:
```suggestion
.xkit-control-button-inner {
```

File: src/util/interface.js [hunk 6]
```diff
@@ -84,5 +84,65 @@
...
+  /**
+   * Create a button template that can be cloned with cloneControlButton() for inserting into the controls of a post.
+   *
+   * @param {string} iconClass The RemixIcon class of the requested icon
+   * @param {string} buttonClass An extra class to identify the extension that added the button
+   * @returns {HTMLDivElement} A button that can be cloned with cloneControlButton()
+   */
+  const createControlButtonTemplate = async function (iconClass, buttonClass) {
+    const controlButtonContainer = document.createElement('div');
+    controlButtonContainer.classList.add('xkit-post-button-container', buttonClass);
```
Comment (AprilSylph) [inline src/util/interface.js:112]:
```suggestion
    controlButtonContainer.classList.add('xkit-control-button-container', buttonClass);
```

File: src/util/interface.js [hunk 7]
```diff
@@ -84,5 +84,65 @@
...
+
+    const controlButtonContainerSpan = document.createElement('span');
+    controlButtonContainerSpan.classList.add('xkit-post-button-container-span');
```
Comment (AprilSylph) [inline src/util/interface.js:115]:
```suggestion
    controlButtonContainerSpan.classList.add('xkit-control-button-container-span');
```

File: src/util/interface.js [hunk 8]
```diff
@@ -84,5 +84,65 @@
...
+
+    const controlButton = document.createElement('button');
+    controlButton.classList.add('xkit-post-button');
```
Comment (AprilSylph) [inline src/util/interface.js:118]:
```suggestion
    controlButton.classList.add('xkit-control-button');
```

File: src/util/interface.js [hunk 9]
```diff
@@ -84,5 +84,65 @@
...
+    controlButton.tabIndex = 0;
+
+    const controlButtonInner = document.createElement('span');
+    controlButtonInner.classList.add('xkit-post-button-inner');
```
Comment (AprilSylph) [inline src/util/interface.js:122]:
```suggestion
    controlButtonInner.classList.add('xkit-control-button-inner');
```

**Review by AprilSylph — CHANGES_REQUESTED** [review]
sorry to reiterate a previous point, but the classnames still use "post button" terminology
(if you bulk-accept these suggestions i think i got all of them)

**Comment by invalidCards** [comment]
That's all of them yeah, thanks

File: src/util/interface.js [hunk 10]
```diff
@@ -84,5 +84,65 @@
     });
   })();
 
-  return { getPostElements, addStyle, removeStyle, registerMeatballItem, unregisterMeatballItem };
+  /**
+   * Fetch the SVG belonging to a specific RemixIcon.
+   *
+   * @param {string} cssClass The RemixIcon class of the requested icon
```
Comment (AprilSylph) [inline src/util/interface.js:90]:
```suggestion
   * @param {string} cssClass - The RemixIcon class of the requested icon
```

File: src/util/interface.js [hunk 11]
```diff
@@ -84,5 +84,65 @@
...
+    if (Object.prototype.hasOwnProperty.call(icons, cssClass)) {
+      return `<path d="${icons[cssClass]}"/>`;
+    }
+  };
+
+  /**
+   * Create a button template that can be cloned with cloneControlButton() for inserting into the controls of a post.
+   *
+   * @param {string} iconClass The RemixIcon class of the requested icon
+   * @param {string} buttonClass An extra class to identify the extension that added the button
```
Comment (AprilSylph) [inline src/util/interface.js:107]:
```suggestion
   * @param {string} iconClass - The RemixIcon class of the requested icon
   * @param {string} buttonClass - An extra class to identify the extension that added the button
```

File: src/util/interface.js [hunk 12]
```diff
@@ -84,5 +84,65 @@
...
+    controlButtonContainer.appendChild(controlButtonContainerSpan);
+
+    return controlButtonContainer;
+  };
+
+  /**
+   * Create a deep-level clone of a button template that is ready to add to the page
+   *
+   * @param {HTMLDivElement} template A button template as returned by createControlButtonTemplate()
+   * @param {Function} callback A function to run when the button is clicked
```
Comment (AprilSylph) [inline src/util/interface.js:138]:
```suggestion
   * @param {HTMLDivElement} template - A button template as returned by createControlButtonTemplate()
   * @param {Function} callback - A function to run when the button is clicked
```

**Review by AprilSylph — CHANGES_REQUESTED** [review]
> for jsdoc `@param` tags, i prefer a hyphen to separate the description from the parameter name

**Review by AprilSylph — CHANGES_REQUESTED** [review]
hopefully my last nitpick here... i feel like these new functions warrant having their own file (`src/util/control_buttons.js`), especially since they have their own stylesheet

File: src/util/control_buttons.js [hunk 13]
```diff
@@ -0,0 +1,63 @@
+(function () {
+  /**
+   * Fetch the SVG belonging to a specific RemixIcon.
+   *
+   * @param {string} cssClass - The RemixIcon class of the requested icon
+   * @returns {string|undefined} The SVG path of the associated RemixIcon if it exists
+   */
+  const getIconPath = async function (cssClass) {
+    const url = browser.runtime.getURL('/src/lib/remixicon-svg.json');
```
Comment (AprilSylph) [inline src/util/control_buttons.js:9]: missed a renamed path

```suggestion
    const url = browser.runtime.getURL('/src/lib/remixicon_svg.json');
```

**Review by AprilSylph — CHANGES_REQUESTED** [review]

**Comment by AprilSylph** [comment]
trying to test this out and `createControlButtonTemplate` results in a `DOMException` on Firefox 82.0b8 (64-bit)... can you reproduce?

File: dev/remix-svg-to-json.js [hunk 14]
```diff
@@ -0,0 +1,23 @@
+const fs = require('fs');
+const parser = require('fast-xml-parser');
+
+const inputPath = './node_modules/remixicon/fonts/remixicon.symbol.svg';
+const outputPath = './src/lib/remixicon_svg.json';
```
Comment (AprilSylph) [inline dev/remix-svg-to-json.js:5]:
```suggestion
const outputPath = './lib/remixicon_svg.json';
```

File: src/util/control_buttons.js [hunk 15]
```diff
@@ -0,0 +1,63 @@
+(function () {
+  /**
+   * Fetch the SVG belonging to a specific RemixIcon.
+   *
+   * @param {string} cssClass - The RemixIcon class of the requested icon
+   * @returns {string|undefined} The SVG path of the associated RemixIcon if it exists
+   */
+  const getIconPath = async function (cssClass) {
+    const url = browser.runtime.getURL('/src/lib/remixicon_svg.json');
```
Comment (AprilSylph) [inline src/util/control_buttons.js:9]:
```suggestion
    const url = browser.runtime.getURL('/lib/remixicon_svg.json');
```

**Review by AprilSylph — CHANGES_REQUESTED** [review]
`src/` and `lib/` are separate folders

**Comment by invalidCards** [comment]
They sure are

**Review by AprilSylph — APPROVED** [review]
tested and LGTM!

here's my working example/test script, which adds a button to posts that logs the post ID to the console when clicked. (might be useful for reference purposes)

```js
(function () {
  let controlButtonTemplate;
  const excludeClass = 'xkit-control-button-test';

  const onButtonClick = ({ target }) => {
    const post = $(target).parents('[data-id]')[0];
    console.log(post.dataset.id);
  };

  const addButtons = async function () {
    const { getPostElements } = await fakeImport('/src/util/interface.js');
    const { cloneControlButton } = await fakeImport('/src/util/control_buttons.js');

    getPostElements({ excludeClass }).forEach(postElement => {
      const controls = postElement.querySelector('footer > :last-child');
      controls.prepend(cloneControlButton(controlButtonTemplate, onButtonClick));
    });
  };

  const main = async function () {
    const { onNewPosts } = await fakeImport('/src/util/mutations.js');
    const { createControlButtonTemplate } = await fakeImport('/src/util/control_buttons.js');

    controlButtonTemplate = await createControlButtonTemplate('ri-terminal-line', 'xkit-control-button-test-button');
    onNewPosts.addListener(addButtons);

    addButtons();
  };

  const clean = async function () {
    const { onNewPosts } = await fakeImport('/src/util/mutations.js');
    onNewPosts.removeListener(addButtons);

    $(`.${excludeClass}`).removeClass(excludeClass);
    $('.xkit-control-button-test-button').remove();
  };

  return { main, clean };
})();
```

**Pull request merged**